File: src/ruby/src/sniper.rb
#encoding: UTF-8

require './src/ruby/util/tester.rb'

def reader(path) File.read(path) end
def writer(path, msg) File.write(path, msg) end

class Sniper
  attr_reader :msg, :state;
  def initialize(items)
    @name, @item, sp, lp, lb, @state = items.split(' ')
    @stop_price, @last_price, @last_bid, @msg = sp.to_i, lp.to_i, lb.to_i, "toAuctionMessage is None"
  end
  def to_s() "#@msg\n#@item #@last_price #@last_bid #@state" end
  def close() @state = @state == "WINNING" ? "WON" : "LOST" end
  def price(price, increment, bidder)
    if bidder == @name then @last_price, @state = price, "WINNING"; return end
    bid = price + increment; if bid > @stop_price then @last_price, @state = price, "LOSING"; return end
    bid_msg = "SOLVersion: 1.1; Command: BID; Price: %d;"
    @last_price, @last_bid, @state, @msg = price, bid, "BIDDING", sprintf(bid_msg, bid)
  end
  def fail(msg) @last_price, @last_bid, @state, @msg = 0, 0, "FAILED", "failed: " + msg + "." end
end

class Translator
  def initialize(msg, sniper) @msg, @sniper, @dic = msg, sniper, {} end
  def mk_hash() @msg.split(';').each{|field| k,v = field.split(':'); k.strip!; v.strip!; @dic[k.to_sym] = v } end
  def get(key) res = @dic[key]; if res == nil then raise "missing field [#{key}]" else res end end
  def event() get(:Event) end; def get_int(key) get(key).to_i end
  def price() get_int(:CurrentPrice) end; def increment() get_int(:Increment) end; def bidder() get(:Bidder) end
  def translate()
    mk_hash;
    @sniper.close if event == "CLOSE"
    @sniper.price(price, increment, bidder) if event == "PRICE"
    rescue => e then @sniper.fail("msg is [#@msg]" + ' ' + e.message)
  end
end

def domain(items, msg)
  sniper = Sniper.new(items)
  trans = Translator.new(msg, sniper)
  trans.translate
  sniper.to_s
end
def str_cut(text, n)
  res = ""
  for i in 0..n do res += text[i] end
  res
end
def test()
  args = "sniper item-123 1300 0 0 JOINING\nSOLVersion: 1.1; Event: CLOSE;"
  res = str_cut(args, 30);
  puts res
end

def refactor()
  items, msg = (reader "a:/pj/sniper/rr/req1.txt").split("\n"); exp = reader "a:/pj/sniper/rr/res2.txt"
  act = domain(items, msg); act = act.gsub("\n", ' '); exp = exp.gsub("\n", ' ')
  zz_act "Sniper Test", act, exp, "此処ではテストが通らない時には詳細がほしい"
end

def developer(args, exp, name)
  act = domain(*args.split("\n")); act = act.gsub("\n", ' '); exp = exp.gsub("\n", ' ')
  zz_act("Sniper Test", act, exp, name)
end
def develop()
  arg1 = reader "a:/pj/sniper/rr/req1.txt"; exp1 = reader "a:/pj/sniper/rr/res1.txt"
  arg2 = reader "a:/pj/sniper/rr/req2.txt"; exp2 = reader "a:/pj/sniper/rr/res2.txt"
  arg3 = reader "a:/pj/sniper/rr/req3.txt"; exp3 = reader "a:/pj/sniper/rr/res3.txt"
  arg4 = reader "a:/pj/sniper/rr/req4.txt"; exp4 = reader "a:/pj/sniper/rr/res4.txt"
  arg5 = reader "a:/pj/sniper/rr/req5.txt"; exp5 = reader "a:/pj/sniper/rr/res5.txt"
  arg6 = reader "a:/pj/sniper/rr/req6.txt"; exp6 = reader "a:/pj/sniper/rr/res6.txt"
  developer arg1, exp1, "参加要請中に限度額以下の価格情報がくると入札中になって買い注文を出すよ"
  developer arg2, exp2, "参加要請中にオークションが閉まると落札は失敗で買い注文は無しだ"
  developer arg3, exp3, "入札中にオークションが閉まっても落札は失敗で買い注文は無しだ"
  developer arg4, exp4, "一位入札中にオークションが閉まれば落札は成功で買い注文は無しだ"
  developer arg5, exp5, "一位入札中でも限度額以上の価格情報がくれば脱落中で買い注文は無しだ"
  developer arg6, exp6, "ヘンテコなメッセージが来ると金額はすべて０にしてエラー発生だ"
end

def product()
  writer("fromSniper.txt", domain(*reader("toSniper.txt").split("\n")))
end

if __FILE__ == $0 then
  # test
  # refactor
  # develop
  product
end
__END__
ruby src/ruby/src/sniper.rb

File: src/python/src/sniper.py
# coding: utf-8
from util.tester import zz_act

def reader(path):
  with open(path) as f: return f.read()
def writer(path, msg):
  with open(path, mode='w') as f: f.write(msg)

class Sniper:
  def __init__(self, items):
    self.name, self.item, sp, lp, lb, self.state = items.split(' ')
    self.stop_price, self.last_price, self.last_bid = int(sp), int(lp), int(lb)
    self.msg = "toAuctionMessage is None"
  def __str__(self): return  f"{self.msg}\n{self.item} {self.last_price} {self.last_bid} {self.state}"
  def close(self): self.state = "WON" if self.state == "WINNING" else "LOST"
  def price(self, price, increment, bidder):
    if bidder == self.name: self.last_price, self.state = price, "WINNING"; return
    bid = price + increment
    if bid > self.stop_price: self.last_price, self.state = price, "LOSING"; return
    self.last_price, self.last_bid, self.state, self.msg = price, bid, "BIDDING", f"SOLVersion: 1.1; Command: BID; Price: {bid};"
  def failed(self, msg):
    self.last_price, self.last_bid, self.state, self.msg = 0, 0, "FAILED", f"failed: {msg}"

class Translator:
  def __init__(self, msg, sniper): self.msg = msg; self.sniper = sniper
  def translate(self):
    dic = {}; fields = self.msg.split(';'); fields.pop()
    for f in fields:
      pair = f.split(':'); dic[pair[0].strip()] = pair[1].strip()
    try:
      event = dic["Event"]
      if (event == "CLOSE"): self.sniper.close(); return
      if (event == "PRICE"): self.sniper.price(int(dic["CurrentPrice"]), int(dic["Increment"]), dic["Bidder"]); return
    except KeyError as msg: self.sniper.failed(f"msg is [{self.msg}] Missing value field is [{str(msg)}]")

def domain(items, msg):
  sniper = Sniper(items); trans = Translator(msg, sniper); trans.translate()
  return str(sniper)
     
def test():
  text = "昔々有るところにおじいさんとおばあさんがおりました。"
  res = ""
  for i in range(10): res += text[i]
  print(res)

def refactor():
  issue = reader("a:/pj/sniper/rr/req1.txt"); exp = reader("a:/pj/sniper/rr/res2.txt")
  items, msg = issue.split('\n'); ans = domain(items, msg); exp = exp.replace('\n', ' '); ans = ans.replace('\n', ' ')
  zz_act("Snier Test", ans, exp, "今回の場合は失敗のメッセージは詳細に表したい")

def developer(args, exp, name):
  items, msg = args.split('\n'); act = domain(items, msg); exp = exp.replace('\n',' '); act = act.replace('\n', ' ')
  zz_act("Sniper Test", act, exp, name)
def develop():
  arg1 = reader("a:/pj/sniper/rr/req1.txt"); exp1 = reader("a:/pj/sniper/rr/res1.txt")
  arg2 = reader("a:/pj/sniper/rr/req2.txt"); exp2 = reader("a:/pj/sniper/rr/res2.txt")
  arg3 = reader("a:/pj/sniper/rr/req3.txt"); exp3 = reader("a:/pj/sniper/rr/res3.txt")
  arg4 = reader("a:/pj/sniper/rr/req4.txt"); exp4 = reader("a:/pj/sniper/rr/res4.txt")
  arg5 = reader("a:/pj/sniper/rr/req5.txt"); exp5 = reader("a:/pj/sniper/rr/res5.txt")
  arg6 = reader("a:/pj/sniper/rr/req6.txt"); exp6 = reader("a:/pj/sniper/rr/res7.txt")
  developer(arg1, exp1, "参加要請中に限度額以下の価格情報がくると入札中になって買い注文を出すよ")
  developer(arg2, exp2, "参加要請中にオークションが閉まると落札は失敗で買い注文は無しだ")
  developer(arg3, exp3, "入札中にオークションが閉まっても落札は失敗で買い注文は無しだ")
  developer(arg4, exp4, "一位入札中にオークションが閉まれば落札は成功で買い注文は無しだ")
  developer(arg5, exp5, "一位入札中でも限度額以上の価格情報がくれば脱落中で買い注文は無しだ")
  developer(arg6, exp6, "ヘンテコなメッセージが来ると金額はすべて０にしてエラー発生だ")

def product():
  items, msg = reader("toSniper.txt").split('\n')
  writer("fromSniper.txt", domain(items, msg))

if __name__ == "__main__":
  # test()
  # refactor()
  develop()
  # product()

'''
python src/python/src/sniper.py
'''